esmvalcore/preprocessor/_multimodel.py

```diff
@@ -201,158 +185,102 @@ def _put_in_cube(template_cube, cube_data, statistic, t_axis):
 
+    This function extracts the date information from the cube and
+    reconstructs the time coordinate, resetting the actual dates to the
+    15th of the month or 1st of july for yearly data (consistent with
+    `regrid_time`), so that there are no mismatches in the time arrays.
 
-def _slice_cube(cube, t_1, t_2):
-    """
-    Efficient slicer.
+    If cubes have different time units, it will use reset the calendar to
+    a default gregorian calendar with unit "days since 1850-01-01".
 
-    Simple cube data slicer on indices
-    of common time-data elements.
+    Might not work for (sub)daily data, because different calendars may have
+    different number of days in the year.
     """
-    time_pts = [t for t in cube.coord('time').points]
-    converted_t = _datetime_to_int_days(cube)
-    idxs = sorted([
-        time_pts.index(ii) for ii, jj in zip(time_pts, converted_t)
-        if t_1 <= jj <= t_2
-    ])
-    return [idxs[0], idxs[-1]]
-
-
-def _monthly_t(cubes):
-    """Rearrange time points for monthly data."""
-    # get original cubes tpoints
-    days = {day for cube in cubes for day in _datetime_to_int_days(cube)}
-    return sorted(days)
-
-
-def _full_time_slice(cubes, ndat, indices, ndatarr, t_idx):
-    """Construct a contiguous collection over time."""
-    for idx_cube, cube in enumerate(cubes):
-        # reset mask
-        ndat.mask = True
-        ndat[indices[idx_cube]] = cube.data
-        if np.ma.is_masked(cube.data):
-            ndat.mask[indices[idx_cube]] = cube.data.mask
-        else:
-            ndat.mask[indices[idx_cube]] = False
-        ndatarr[idx_cube] = ndat[t_idx]
-
-    # return time slice
-    return ndatarr
-
+    t_unit = _get_consistent_time_unit(cubes)
 
-def _assemble_overlap_data(cubes, interval, statistic):
-    """Get statistical data in iris cubes for OVERLAP."""
-    start, stop = interval
-    sl_1, sl_2 = _slice_cube(cubes[0], start, stop)
-    stats_dats = np.ma.zeros(cubes[0].data[sl_1:sl_2 + 1].shape)
-
-    # keep this outside the following loop
-    # this speeds up the code by a factor of 15
-    indices = [_slice_cube(cube, start, stop) for cube in cubes]
-
-    for i in range(stats_dats.shape[0]):
-        time_data = [
-            cube.data[indx[0]:indx[1] + 1][i]
-            for cube, indx in zip(cubes, indices)
-        ]
-        stats_dats[i] = _compute_statistic(time_data, statistic)
-    stats_cube = _put_in_cube(cubes[0][sl_1:sl_2 + 1],
-                              stats_dats,
-                              statistic,
-                              t_axis=None)
-    return stats_cube
-
-
-def _assemble_full_data(cubes, statistic):
-    """Get statistical data in iris cubes for FULL."""
-    # all times, new MONTHLY data time axis
-    time_axis = [float(fl) for fl in _monthly_t(cubes)]
+    for cube in cubes:
+        # Extract date info from cube
+        years = [cell.point.year for cell in cube.coord('time').cells()]
```

Comment:
```suggestion
        coord = cube.coord('time')
        years = [p.year for p in coord.units.num2date(coord.points)]
```
using the function [`cftime.num2date`](https://unidata.github.io/cftime/api.html#cftime.num2date) on a numpy array appears to be a lot faster than using the `cells()` method of the coordinate to iterate over the cells, also in other places